### ISSUE TITLE
DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead

### ISSUE THREAD

**Issue by cosminn777** [Automattic/mongoose#5399]
DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`

Mongoose 4.11.0, MongoDB 2.2.29, NodeJS 8.1.2

**Comment by alanpurple** [comment]
+1  , I have no idea where to fix to get rid of that warning

**Comment by tinovyatkin** [comment]
The warning is for a code inside Mongoose actually, when connecting string has a replica set:

```js
Mongoose.prototype.connect = function() {
  var conn = this.connection;
  if ((arguments.length === 2 || arguments.length === 3) &&
      typeof arguments[0] === 'string' &&
      typeof arguments[1] === 'object' &&
      arguments[1].useMongoClient === true) {
    return conn.openUri(arguments[0], arguments[1], arguments[2]);
  }
  if (rgxReplSet.test(arguments[0]) || checkReplicaSetInUri(arguments[0])) {
    return new MongooseThenable(this, conn.openSet.apply(conn, arguments));
  }

  return new MongooseThenable(this, conn.open.apply(conn, arguments));
};
```

**Comment by alanpurple** [comment]
@tinovyatkin  

So then, this is a bug?

**Comment by tinovyatkin** [comment]
looks like a bug, can we workaround by adding `useMongoClient: true` to options (you will see a deprecation from MongoDB driver then but that doesn't throw)

**Comment by matthiashagemann** [comment]
Adding `useMongoClient: true` makes the message disappear, but then my documents simply stop loading. Haven’t debugged it deeper, but I would prefer to hear best practices first before I adapt my code.

**Comment by artnicbr** [comment]
My application also doesn't work with the workaround suggested above. It simply doesn't run MyModel.find method anymore and gives no error and no timeout.

**Comment by alanpurple** [comment]
The reason why "useMongoclient: true" option doesn't work for most of the cases is for that option, every mongoose connection-related object returns dummy object cause not able to access mongodb

I tried to use it like

```
const server = express();
mongoose.connect('mongodb://localhost/advisorDemoTestDB', { useMongoClient: false })
    .then(() => require('./db-init')(server))
    .catch(err => console.error(err));
```

but it doesn't work

**Comment by JCMais** [comment]
Having same issue, even without replica sets.

**Comment by bertolo1988** [comment]
Having same issue:

```
(node:4138) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`
```

**Comment by DZuz14** [comment]
Same...

**Comment by z1000zhihe** [comment]
After I added below code to my mongo connection, I can't query out anything now.
{useMongoClient: true}.
Any suggestions will be appreciate!!

**Comment by angeloLed** [comment]
+1 same with replica sets, no sharding

**Comment by niftytushar** [comment]
+1

**Comment by evandersar** [comment]
+1

**Comment by prawee** [comment]
+1

**Comment by yduman** [comment]
+1 my build is failing because of that

**Comment by svanlangen** [comment]
+1

**Comment by Yurockkk** [comment]
+1

**Comment by IoHL** [comment]
+1

**Comment by Math3v** [comment]
+1

**Comment by Chenz62** [comment]
The easiest fix for this;  "npm remove mongoose" then "npm install mongoose@4.6.0 --save" problem solved. Upgrading is not always the best option.

**Comment by toddmurphy** [comment]
+1

**Comment by rodrigoalviani** [comment]
+1

**Comment by Systerr** [comment]
+1

**Comment by EricRabil** [comment]
+1

**Comment by plhosk** [comment]
I get 2 errors: 
```javascript
(node:9260) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`
Server started on port 3000
Db.prototype.authenticate method will no longer be available in the next major release 3.x as MongoDB 3.6 will only allow auth against users in the admin db and will no longer allow multiple credentials on a socket. Please authenticate using MongoClient.connect with auth credentials.
```
The first error appeared in 4.11.0. The second error also appeared in the previous version of Mongoose.

**Comment by cosminn777** [comment]
plhosk, the second one should have been fixed in 4.11 but looks like it's still there, for me too.

**Comment by plhosk** [comment]
@cosminn777, I guess the `useMongoClient = true` option was added as the way to get rid of the 2nd error, but it has a bug that causes Mongoose to stop working.

**Comment by jlobos** [comment]
+1

**Comment by Alhern** [comment]
+1 sadly.

**Comment by DellAnderson** [comment]
Installed 4.10.8 without issues.  Please consider changing `npm install mongoose --save` to default to 4.10.8 until 11 is stable.

**Comment by nicholaai** [comment]
+1

**Comment by montoias** [comment]
+1

**Comment by NguyenManh94** [comment]
+1

**Comment by mattmengSemioe** [comment]
+1

**Comment by itpcc** [comment]
I also found this issue too.

__My MongoDB__ emvironment:
```log
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten] db version v3.4.5
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten] git version: 520b8f3092c48d934f0cd78ab5f40fe594f96863
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.0.1t  3 May 2016
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten] allocator: tcmalloc
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten] modules: none
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten] build environment:
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten]     distmod: debian81
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten]     distarch: x86_64
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten]     target_arch: x86_64
2017-06-20T08:04:24.509+0000 I CONTROL  [initandlisten] options: { security: { authorization: "enabled" } }
```

My connection configuration:
```js
var dbURL = `mongodb://${dbHost}:${dbPort}/${dbName}?authSource=admin`;
var dbAuth = { 
	useMongoClient: false
	user: dbUser,
	pass: dbPass
}
mongoose.connect(dbURL, dbAuth);
```

When I use `useMongoClient` to be false. Mongoose show following warnings:
```log
(node:7868) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`
Express server listening on port 3222 in development mode
Db.prototype.authenticate method will no longer be available in the next major release 3.x as MongoDB 3.6 will only allow auth against users in the admin db and will no longer allow multiple credentials on a socket. Please authenticate u
sing MongoClient.connect with auth credentials.
```

But it work just fine.

However, when I set it to `true`, I found this error:
```log
Unhandled rejection MongoError: not authorized on users to execute command { listIndexes: "sessions", cursor: {} }
    at Function.MongoError.create (C:\Users\itpcc\Desktop\gable\project\kep-user\app\node_modules\mongoose\node_modules\mongodb-core\lib\error.js:31:11)
    at queryCallback (C:\Users\itpcc\Desktop\gable\project\kep-user\app\node_modules\mongoose\node_modules\mongodb-core\lib\cursor.js:212:36)
    at C:\Users\itpcc\Desktop\gable\project\kep-user\app\node_modules\mongoose\node_modules\mongodb-core\lib\connection\pool.js:469:18
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickCallback (internal/process/next_tick.js:98:9)
```

**Comment by ajaymore** [comment]
(node:451) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use
`openUri()` instead, or set the `useMongoClient` option if using `connect()` or `
createConnection()`                                                              
Db.prototype.authenticate method will no longer be available in the next major re
lease 3.x as MongoDB 3.6 will only allow auth against users in the admin db and w
ill no longer allow multiple credentials on a socket. Please authenticate using M
ongoClient.connect with auth credentials.


I get this issue with mongoose version 4.11.0 and MongoDB version 3.4.5

**Comment by jcurlier** [comment]
+1

**Comment by jlanio** [comment]
+1

**Comment by nandofalcao** [comment]
Use
```
mongoose.createConnection(URI)
```

**Comment by jlanio** [comment]
@nandofalcao When using mongoose.createConnection (URI) or mongoose (URI, {useMongoClient: true})
Apparently I can not save new records.

```
let userSchema = mongoose.Schema({ name: String });
let User = mongoose.model('User', userSchema);
let joao = new User({ name: 'NewJoao ' });
joao.save((err) => err ? throw err : console.log('User created!'));//nothing happens
```

**Comment by amackintosh** [comment]
Adding polyglot tonight and +1 experiencing this.

```
// Load Mongoose
const mongoose = require('mongoose');
// import mongoose from 'mongoose' didn't give access to .connect()

// Use native Promises
mongoose.Promise = global.Promise;

// Connect database
export const Mongoose = new Promise((resolve, reject) => {
  const uri = `mongodb://${Singleton.currentConfig.databases.mongodb.host}/${Singleton.currentConfig.databases.mongodb.database}`;

  const options = {
    user: Singleton.currentConfig.databases.mongodb.user,
    pass: Singleton.currentConfig.databases.mongodb.password,
    server: {
      reconnectTries: Singleton.currentConfig.databases.mongodb.reconnectTries,
      reconnectInterval: Singleton.currentConfig.databases.mongodb.reconnectInterval,
      socketOptions: {
        keepAlive: Singleton.currentConfig.databases.mongodb.keepAlive,
        connectTimeoutMS: Singleton.currentConfig.databases.mongodb.connectTimeoutMS
      },
    },
  };

  // Initiate document store
  mongoose.connect(uri, options)

  // Check for anomalies
  .then((connected) => {
    if (mongoose.connection.readyState !== 1) {
      reject(connected);
    }
    resolve(connected);
  })

  // Complete meltdown
  .catch((error) => {
    console.log(`MongoDB Connection Error: ${error}`);
    process.exit(0);
  });
});

```

**Comment by x1911** [comment]
+1

and i get the 2nd message:


[2017-06-27 16:14:23.702] [INFO] :: - Server started in  port 2000  

(node:1193) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`

Db.prototype.authenticate method will no longer be available in the next major release 3.x as MongoDB 3.6 will only allow auth against users in the admin db and will no longer allow multiple credentials on a socket. Please authenticate using MongoClient.connect with auth credentials.

**Comment by skal88** [comment]
+1

**Comment by galaxycubic** [comment]
+1

**Comment by shaunfrisbee** [comment]
+1

**Comment by greyyoung** [comment]
+1

**Comment by lukasrakauskas** [comment]
+1

**Comment by Sipike** [comment]
+1

**Comment by aluxian** [comment]
Can somebody lock this to prevent useless +1's?

**Comment by matthiashagemann** [comment]
As a courtesy to subscribers of this thread, please **stop** replying with +1 because it generates useless email notifications. This is a not a forum. If you want to follow this thread, use the **subscribe** button on the right sidebar. You could also do some research yourself and contribute solutions.

**Comment by z1000zhihe** [comment]
I changed version back to 4.10.8 and it's working fine now.

**Comment by phased90** [comment]
Looks like ValYouW has a a potential fix to the document loading problem here: #5404

**Comment by Smiley01987** [comment]
+1

**Comment by crisamdegracia** [comment]
mongoose.connection.openUri('mongodb://127.0.0.1/camp_v12')

has anyone tried this? my deprecated warning disappeared when i use this, it was from the documentation

http://mongoosejs.com/docs/connections.html

**Comment by simonemazzoni** [comment]
@crisamdegracia that assumes you already has a connection

```
// Or, if you already have a connection
connection.openUri('mongodb://localhost/myapp', { /* options */ });
```

which we don't (at least me since I used to connect to mongodb using mongoose.connect method)

**Comment by mahnunchik** [comment]
```
mongoose.connect(config.get('mongo'), {
  useMongoClient: true,
});
```

I could not find any document... Whats wrong?

**Comment by ewan777** [comment]
+1

**Comment by jbenner-radham** [comment]
So after looking at the linked issue from @phased90 that explains that you can connect like so:

```js
const mongoose = require('mongoose');

mongoose.Promise = global.Promise;
mongoose.connect('mongodb://localhost/test', {useMongoClient: true})
	.then(() => {
		let Cat = mongoose.model('Cat', {name: String});
		let kitty = new Cat({name: 'Zildjian'});

		kitty.save(err => {
			if (err) {
				console.log(err);
			} else {
				console.log('meow');
			}
		});
	})
	.catch(err => console.error(err));
```

I did a quick dive through the code and it appears that 4.11 returns `MongooseThenable` except for when the `{useMongoClient: true}` option is provided. In that instance it calls `Connection.prototype.openUri` which returns a new instance of `Promise.ES6`. Should that `Connection.prototype.openUri` call get wrapped in a `MongooseThenable`?

My apologies for the question, this is my first time looking at this codebase so I'm a bit unsure. But I would be happy to try and help provide a fix if I can get some clarification on if `Moongoose.prototype.connect` should always be returning an instance of `MongooseThenable`.

**Comment by agoldis** [comment]
Be aware that using `{useMongoClient: true}` will cause `Connection.prototype.openUri` to be called, which doesn't trigger buffered commands that were created __before__ connecting to a DB due to #5404

**Comment by adrianodias** [comment]
+1

**Comment by ivanhuay** [comment]
+1

**Comment by varunjayaraman** [comment]
Sorry, things have been crazy, will take a look in a bit.

**Comment by bertolo1988** [comment]
@varunjayaraman no need to be sorry mate. People can either wait, or try to fix it by themselves.

**Comment by serganus** [comment]
+1

**Comment by varunjayaraman** [comment]
So regarding the actual issue that was opened, the deprecation warning is not a bug, it's just a warning: mongoose is going to deprecate its current connection logic, so you should pass in the `useMongoClient` param. This can be found in the [documentation](http://mongoosejs.com/docs/connections.html)

On a separate note, it looks like this was a backwards breaking change, hopefully this can be fixed in the next minor release (or maybe it isn't, I'm running on very little sleep here):

this would normally work (save for the `useMongoClient` part, which is new):

```javascript
const mongoose = require('mongoose');
const co = require('co');
mongoose.Promise = global.Promise;
const GITHUB_ISSUE = `gh-5399`


exec()
  .then(() => {
    console.log('successfully ran program');
    process.exit(0);
  })
  .catch(error => {
    console.error(`Error: ${ error }\n${ error.stack }`);
  });


function exec() {
  return co(function*() {
    const db = mongoose.createConnection(`mongodb://localhost:27017/${ GITHUB_ISSUE }`, { useMongoClient: true })
    const schema = new mongoose.Schema({
      name: String
    });


    const Model = db.model('Model', schema);
  });
}
```

This seems to throw an error for me because `db.model` is no longer a function.

When using `mongoose.model`, it works:

```javascript
const mongoose = require('mongoose');
const co = require('co');
mongoose.Promise = global.Promise;
const GITHUB_ISSUE = `gh-5399`


exec()
  .then(() => {
    console.log('successfully ran program');
    process.exit(0);
  })
  .catch(error => {
    console.error(`Error: ${error}\n${error.stack}`);
  });


function exec() {
  return co(function* () {
    const db = mongoose.connect(`mongodb://localhost:27017/${GITHUB_ISSUE}`, { useMongoClient: true })

    return db
      .then(() => {
        const schema = new mongoose.Schema({
          name: String
        });


        const Model = mongoose.model('Model', schema);

        return Model.create({ name: 'test' });
      })
      .then(doc => console.log(doc));

  });
}
```

So it looks like there is something up with the `thenable`

**Comment by itpcc** [comment]
@varunjayaraman I've just update Mongoose and use `useMongoClient: true` as you suggest, but I have an issue.
It doesn't support, according to [the official Mongoose documentation](http://mongoosejs.com/docs/connections.html), `user`,  and `pass`  options:

```log
the options [user] is not supported
the options [pass] is not supported
```

And, yeah, failed to authen. :-1:

**Comment by CQBinh** [comment]
@itpcc try to use `4.10.8` version is the best solution at this time.
```
npm remove mongoose
npm install mongoose@4.10.8 --save
```
Credit: @Chenz62

**Comment by itpcc** [comment]
@CQBinh Thank you for your advice. I already read those comments about it. 
However, I don't think using the older version is the best solution for now. And it can resolve by insert into connection string.
I just feel confused because it was mentioned in the document but it doesn't work. :/
BTW, I report as a new issue [here](https://github.com/Automattic/mongoose/issues/5419)

**Comment by danieljuhl** [comment]
@varunjayaraman if `db.model` is no longer present, how would one handle multiple different mongoose connection with differente models?

eg model1 and model2 attached to mongooseConnection1 and model3 and model4 attached to mongooseConnection2.

Previously this was done by calling `mongoose.createConnection` for each of the connections, and attaching the models to each of these return values.

**Comment by boyce-ywr** [comment]
+1

**Comment by simonemazzoni** [comment]
@boyce-ywr SERIOUSLY, another +1??? What's wrong with you people??
Post constructive comments, not stuff like that.

Regarding mongoose, it's not clear to me if the `mongoose.connect()` method will be replaced in future releases or not. Someone could explain please?

**Comment by Chenz62** [comment]
@simonemazzoni You are right, they're SPAMming us.

### If you want to be included in the notifications for this thread, all you need to do is subscribe to it (there's a button at the top of the thread), 

All your doing when you **+1** is send everyone here an **email** saying 

> Hello I'm here!

We don't care, we're waiting for a constuctive comments and a solution.

**Comment by MyBoon** [comment]
Like some people said before, the only "solution" is to go back to _mongoose@4.10.8_ ...
At least for now!

**Comment by runners4tme** [comment]
Looks like a bug, went back to mongoose@4.10.8 to solve the problem

**Comment by diegoazh** [comment]
Use mongoose.createConnection(...) instead of mongoose.connect(...)

**Comment by midnightcodr** [comment]
Here's a complete working example of using mongoose@4.11.0 based on @diegoazh 's recommendation
```javascript
const mongoose = require('mongoose')
mongoose.Promise = require('bluebird')  // optional, use this to get rid of 
                                        // the mpromise DeprecationWarning
const conn = mongoose.createConnection('mongodb://localhost/testDB')
const Schema = mongoose.Schema

const UserSchema = new Schema({
	username: String,
	email: String
})

const User = conn.model('User', UserSchema)
module.exports = User

```

**Comment by bertolo1988** [comment]
@midnightcodr So we just drop connect and use createConnection instead?

**Comment by midnightcodr** [comment]
@bertolo1988 just like ppl suggested on this issue, there are currently two workarounds
1. Fall back to 4.10.8
2. Use the new pattern
```javascript
const conn = mongoose.createConnection(...)
const Model = conn.model(...)
```
I personally prefer the second solution.

**Comment by alanpurple** [comment]
Thanks!!! 
solved by createConnection, and use that connection to make model ( instead of mongoose )

```
let mongoose = require('mongoose');

mongoose.Promise = global.Promise;

module.exports = mongoose;
```

```
const mongoose = require('./set-mongoose');
module.exports = mongoose.createConnection('mongodb://localhost/advisorDemoTestDB');
```

**Comment by cosminn777** [comment]
The new connection pattern requires some code changes, and depending on the size of your project that can mean a lot of time and maybe even architecture changes (given the fact that creating a model requires a connection).

It would have been great if the old connection model could wrap the new connection model.

**Comment by peterpeterparker** [comment]
There is a lot of post, so may I try to summarize:

1. The warning is a deprecated warning, meaning, there is no rush. If we don't change anything it still gonna work as it always has, everything seems to be fine

2. According the warning, connect() gonna be/has been modified and passing useMongoClient gonna be a must. But here is the possible bug/problem

- connect('myurl', {useMongoClient: true}) => the document stop loading and it's not possible to write in the db

- connect('myurl', {useMongoClient: false}) => seems to work fine but the warning is still displayed

Therefore the purpose of this issue is to correct the use of useMongoClient in case of the call of the method connect.

Agree?

**Comment by Archomeda** [comment]
Still, while a deprecation warning is indeed "just" warning, having that show up makes me wonder when it's actually going to be removed. If mongoose is following semver, it should still be there in the 4.x branch. Therefore, it should still work with backwards compatibility until 5.x is released.

Regarding the change itself, it's kind of annoying that I couldn't quickly find anything regarding this change except the one mentioned in the changelog. And even when you do change it, stuff still breaks and according to multiple people here, it's because of the `.model()` call.

Also, even the docs still show `mongoose.connect()` and `mongoose.model()`, which makes this matter even more confusing. The `mongoose.connect()` change is only mentioned under Connection pools, and not on other parts of [this page](http://mongoosejs.com/docs/connections.html). So the question remains, is this change a bug, or intended?

**Comment by mkastner** [comment]
I don't know where this is going, but I've spent the better part of this weekend trying get a connection going to MongoDB's Atlas. The native mongodb driver was instantly up and running the CRUD stuff.

However trying to get mongoose going is a total nightmare. If I use createConnection with 4.11 the cluster connections get killed:

    MongoError: connection 4 to cluster0-myshard.mongodb.net:27017 closed

Alternatively, I could set 

    useMongoClient: true

but then I get undefined on the connection.model

    <info> atlas-api-test.js:65 (Test.testVendors) conn.model undefined

**Comment by vkarpov15** [comment]
@Archomeda no plans to remove legacy connection behavior in 4.x, and 5.x is blocked on getting the new connection logic to a steady state. There are docs on the [connection docs](http://mongoosejs.com/docs/connections.html) that 4.11.1 will improve and link to that should be helpful. Re: the `model()` call, if you specify `useMongoClient: true`, `createConnection()` will **not** return a connection, but rather a promise that resolves to a connection that you need to `.then()` on. Rationale is described here: https://github.com/Automattic/mongoose/issues/5404#issuecomment-312522539 Please feel free to voice any concerns about this on issue #5404 .

@mkastner looks like you're using `createConnection()`. With `useMongoClient: true`, `createConnection()` returns a promise that resolves to a connection. Please use `await mongoose.createConnection()` or `mongoose.createConnection().then()`

@peterpeterparker #5404 pointed out an issue with the new connection logic that will be fixed in 4.11.1, to be released within the next 24 hours.

@cosminn777 what architectural issues are you having? Please open up a separate issue to discuss them. The long-term plan is that in 5.x we will not support Node < 4, so everyone using mongoose will have native access to generators and/or async/await, so this change would be as easy as adding a `yield` or `await` statement.

@alanpurple the issue that you're referring to is the same as the one mentioned #5404 and will be fixed with 4.11.1.

@itpcc I opened up a separate issue to track #5432. For now, please put your username/password in the URI: `var dbURL = `mongodb://${dbUser}:${dbPass}@${dbHost}:${dbPort}/${dbName}?authSource=admin`.

In general, thanks everyone for your patience in trying this new behavior. I realize that switching over will cause some headache, which is why we hid this behavior behind a flag, but some changes coming down the pipe for MongoDB 3.6 necessitate a complete overhaul of mongoose's internal connection logic. I hope the new connection logic will lead to fewer mongoose-specific connection bugs in the long run, but please bear with us as we iron out the behavior and documentation. Your feedback is much appreciated.

As long as you're on 4.x your current connection logic will work, it will just print a warning. In the meantime, please feel free to report any issues you have in a separate github issue.

**Comment by mathewmorris** [comment]
Just adopted the new method and it works great. Thanks everyone!

**Comment by adamreisnz** [comment]
@vkarpov15 thanks for the clarifications. 

As a FYI, the same deprecation warning is thrown with `openSet` as well (I am using a replica set).

So to summarise, it is safe to ignore this warning for now as it only relates to Mongoose's internal workings, and we could try to use `useMongoClient` but don't have to yet.

How stable is it do you reckon? Is it safe to use? I wouldn't want to spend 2 hours refactoring all our models and connection logic to find out that it's unusable in production still.

Is it possible to perhaps remove the deprecation warning until such a time that using the flag is stable and the recommended way to connect going forward? Or has that point been reached already?

**Comment by mkastner** [comment]
@vkarpov15

But I am using async on createConnection. I've set up two tests against the db server. The first one is the native mongo-js canary test to make sure the url works and I can write and read:

    tape('test connection via mongo native', async function testMongoConnection(t) {

      try {
    
        t.plan(1);
        let db = await MongoClient.connect(dbURL);
        await db.collection('users').insert({name: 'username'});
        let result = await db.collection('users').find({});
        let docs = await result.toArray();
        log.info('result', await result.toArray());
        t.equal(1, docs.length);
        db.collection('users').remove({});

        await db.close();

      } catch (err) {
      console.error(err)
      }
    });

Result:

    2017-07-04T09:00:34+0200 <info> atlas-api-test.js:28 (Test.testMongoConnection) result [ { _id: 595b3d1146734207bad88f9d, name: 'username' } ]
    ✔ should be equal

Succeeded by the mongoose test using mongoose createConnection with await:

    tape('test connection via mongoose', async function testMongooseConnection(t) {

      try {
          let conn = await mongoose.createConnection(dbURL, {
            useMongoClient: true,
            /*
            // tried these settings too
            user: 'myUserName',
            pass: 'myPassword',
            connectTimeoutMS: 30000,
            rs_name: 'myReplSetName',
            ssl: true,
            auto_reconnect: true,
            */
          });

          let UserSchema = new mongoose.Schema({
            name: String,
          });

          let User = mongoose.model('User', UserSchema);
          let newUser = new User({name: 'username'});
          let createdUser = await newUser.save();

          // just ending test without validating any results
          t.end();

      } catch (err) {
        log.error(err);
      }

    });

And it throws the following error:

    /mypath/node_modules/mongodb/lib/replset.js:390 process.nextTick(function() { throw err; })
                                        ^
    MongoError: connection 8 to cluster0-shard-00-02-c4nst.domain.com:27017 closed at Function.MongoError.create (/mypath/node_modules/mongodb-core/lib/error.js:29:11)
    at Socket.<anonymous> (/mypath/node_modules/mongodb-core/lib/connection/connection.js:202:22)
    at Object.onceWrapper (events.js:316:30)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at TCP._handle.close [as _onclose] (net.js:549:12)`

I might be terribly wrong here, but to me it doesn't look like an async/promise problem;

**Comment by Thinkdiff** [comment]
I`m using Mongo Atlas. When I use {useMongoClient: true}, My app will connect to the database successfully but the database does not a response to the application, way?

My code:
`mongoose.connect(config.dbhost,  {useMongoClient: true}, function(err){
    if(err){
        console.log(err);
    } else {
        console.log('connected to the database successfuly.');
    }
});`

**Comment by aluxian** [comment]
Just a notice for everyone because I haven't seen it mentioned here: 4.11.1 has been released and after a quick look it seems the buffering problem when using useMongoClient=true has been fixed. I'll test more tomorrow. Anyone else?

**Comment by saostad** [comment]
What is the final solution?

**Comment by Blackbaud-SteveBrush** [comment]
This worked for me:

**database.js**
```
mongoose.connect(databaseUri, { useMongoClient: true })
      .then(() => logger.info(`Database connected at ${databaseUri}`))
      .catch(err => logger.error(`Database connection error: ${err.message}`));
```

**user.model.js**
```
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const userSchema = new Schema({ ... });
const User = mongoose.connection.model('User', userSchema);
```

**Comment by peterpeterparker** [comment]
Could someone explain briefly what exactly is { useMongoClient: true } for or what does that flag do? 

I'm a little bit lost,  what's the advantage of using it? 

I mean, like I said before, I could live with the warning therefore I don't feel the rush to add this param.
Didn't find that much in the doc, thx in advance for clearing that up.

**Comment by Thinkdiff** [comment]
@Blackbaud-SteveBrush It does not work. I`m using Mongo Atlas. Why It does not work?

**Comment by kyrylkov** [comment]
@Thinkdiff 

Doesn't work how?  It works for me on Atlas with usual:

`const User = mongoose.model('User', userSchema);`

instead of 

`const User = mongoose.connection.model('User', userSchema);

**Comment by real-artswan** [comment]
If I try to enable useMongoClient option with few hosts defined in the URI, It hangs and then throws error: 
`MongoError: no mongos proxy available at Timeout._onTimeout (/home/ubuntu/code/pss/node_modules/mongoose/node_modules/mongodb-core/lib/topologies/mongos.js:636:28) at ontimeout (timers.js:386:14) at tryOnTimeout (timers.js:250:5) at Timer.listOnTimeout (timers.js:214:5) name: 'MongoError', message: 'no mongos proxy available'`
uri: mongodb://10.0.1.49:27017,10.0.2.158:27017,10.0.3.84:27017/pss
options: { useMongoClient: true, autoReconnect: true, keepAlive: 300000, connectTimeoutMS: 30000 }

**Comment by Blackbaud-SteveBrush** [comment]
For anyone interested, here are the Mongoose docs on the `useMongoClient` parameter: http://mongoosejs.com/docs/connections.html#use-mongo-client

"Mongoose's default connection logic is deprecated as of 4.11.0...This deprecation is because the MongoDB driver has deprecated an API that is critical to mongoose's connection logic to support MongoDB 3.6, see this github issue for more details." Here's the GitHub issue: https://github.com/Automattic/mongoose/issues/5304

**Comment by peterpeterparker** [comment]
@Blackbaud-SteveBrush thx for pointing that our. So if I understand correctly

useMongoClient = true => activate new connection logic which will also work with Mongo 3.6
useMongoClient = false => default value. old connection logic < 4.11.0. the one which is deprecated.

right?

**Comment by peterpeterparker** [comment]
@Blackbaud-SteveBrush about your piece of code, in my case I don't use

    const User = mongoose.connection.model('User', userSchema);

but 

    const User = mongoose.model('User', userSchema);

and it was still successful. thx

**Comment by vale314** [comment]
Mongoose works for me 4.10.8 :|

**Comment by narendeveloper1** [comment]
I've found that this is likely a bug with mongoose

problem went away after rolling back mongoose version

npm uninstall -save mongoose
npm install -save mongoose@4.10.8

**Comment by yuricamara** [comment]
I have completely ignored this warning and everything is working as expected. Maybe, this the solution for now. Just ignore.

**Comment by daviskoko** [comment]
rolling back to 4.10.8 got my connection working again however i still get the error

`Db.prototype.authenticate method will no longe against users in the admin db and will no long connect with auth credentials.`

**Comment by vkarpov15** [comment]
@peterpeterparker correct. Mongoose's current auth code will not work on mongodb 3.6 based on my understanding, so `useMongoClient` represents us trying to get ahead of this problem while maintaining backwards compatibility. As far as I know there is no scheduled release date for mongodb 3.6 yet, so switching over is not pressing.

@adamreisnz we've had a few bugs but really the new connection logic is a very thin wrapper around using the native driver's MongoClient.connect() function, which has been the preferred method for connecting to mongodb for a couple years now. Once your connection is established its the same mongodb driver logic maintaining the connection. `useMongoClient` should only affect initial connection.

**Comment by peterpeterparker** [comment]
@vkarpov15 thx a lot for the explanation and of course for the work, really appreciate it!

**Comment by adamreisnz** [comment]
@vkarpov15 thanks, I'll give it a go and switch to using `useMongoClient` and see where it takes us.

**Comment by yuen1318** [comment]
const mongoose = require("mongoose");
 const db = "mongodb://localhost/testaroo";

 mongoose.connection.openUri(db);
 mongoose.connection.once("open", function () {
     console.log("connection established");
 }).on("error", function (error) {
     console.log(error);
 })



--i use this and no errors found

**Comment by vitorbarros** [comment]
```
var mongoose = require('mongoose');
mongoose.Promise = global.Promise;

var mongoDB = mongoose.createConnection('mongodb://your_database', {
    useMongoClient: true
});

mongoDB
    .then(function (db) {
        console.log('mongodb has been connected');
    })
    .catch(function (err) {
        console.log('error while trying to connect with mongodb');
    });

module.exports = mongoDB;
```

I use this and no errors found

**Comment by lowato** [comment]
@vitorbarros It worked for me. Thanks!

**Comment by saostad** [comment]
It works for me:
`http://mongoosejs.com/docs/connections.html#use-mongo-client`

**Comment by lgomez** [comment]
I ran this quick test and it worked:
```javascript
// index.js
const mongoose = require('mongoose')
mongoose.Promise = global.Promise;
const db = mongoose.createConnection(`mongodb://localhost/test`)

db.on('error', err => console.log(err))
db.once('open', () => {
  console.log(`Connected to Mongo at: ${new Date()}`)
  db.close(() => {
    console.log(`Disconnected from Mongo at: ${new Date()}`)
  })
})
```
then `node index.js` give me:
```
Connected to Mongo at: Thu Jul 13 2017 22:54:50 GMT+0000 (UTC)
Disconnected from Mongo at: Thu Jul 13 2017 22:54:50 GMT+0000 (UTC)
```
Thank you all!

**Comment by feujoseiii** [comment]
@vitorbarros your solution works! thanks.

**Comment by adamreisnz** [comment]
I can confirm that with 4.11.3 replacing the connection logic with what @vitorbarros suggested removes the warnings and seems to work well 👍 

There is no need to refactor model references as suggested earlier here in this thread.

**Comment by boaz-amit** [comment]
@vitorbarros's solution is indeed the recommended way according to ["The `useMongoClient` Option" documentation](http://mongoosejs.com/docs/connections.html#use-mongo-client).

**However, note that `mongoose.Promise = global.Promise;` will _not_ promisify the underlying Mongo driver**. According to the ["Promises for the MongoDB Driver" documentation](http://mongoosejs.com/docs/promises.html#promises-for-the-mongodb-driver), the `promiseLibrary` property should also be set in the options object of the `connect` method.

For example:
```
var mongoDB = mongoose.connect('mongodb://your_database', {
    useMongoClient: true,
    promiseLibrary: global.Promise
});
```
`global.Promise` would use the native promise engine. Naturally, it's also possible to use Bluebird or any other engine, e.g. `promiseLibrary: require('bluebird')`

**Comment by aprisniak** [comment]
@vitorbarros solution works for me. And also we need to use promiseLibrary like @boaz-amit said. 
But which way is the best for using authentication data? In URL 
`mongodb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]` or better set user and password in options? I can't find right place for now.

**Comment by adamreisnz** [comment]
It has to be in url, the user / pass options no longer work with
useMongoClient true I found.

**Comment by vkarpov15** [comment]
@adamreisnz @aprisniak  working on it, see #5419. In the meantime, please put the username and password in your URI: `mongodb://user:pass@hostname:port/db`

Also, I wrote a blog post about this option and why it is necessary: http://thecodebarbarian.com/mongoose-4.11-use-mongo-client.html

**Comment by iamdubx** [comment]
@vkarpov15 I am using user:pass in URI, but still have the warning.

**Comment by gangstara2** [comment]
For those who use mongoose with gridfs-stream:
`mongoose.Promise = global.Promise
const mongooseConnect = mongoose.connect(
	process.env.MONGODB_URI || process.env.MONGOLAB_URI,
	{
		useMongoClient: true
	}
)

let gfs
mongooseConnect.then(conn => {
	gfs = Grid(conn.db)
	/** Setting up GFS storage here */
})
`

**Comment by FinalDes** [comment]
add the missing useMongoClient option in typescript typings.

**Comment by iamdubx** [comment]
I am not sure should I do something with this warning or not? It will be fixed, right?

**Comment by FinalDes** [comment]
I just add the useMongoClient in typings to make it work. @iamdubx

**Comment by iamdubx** [comment]
@FinalDes how? I added `mongoose.connect(process.env.DATABASE, { useMongoClient: true })` but I still got the message. And I am using user:pass in URI

**Comment by Michael-Abebe** [comment]
Better way to get rid of such problem is to use these two commands:
npm uninstall -save mongoose
npm install -save mongoose@4.10.8

NB: mongoose@4.10.8 is the stable version.

**Comment by Snow-Okami** [comment]
How are you all handling your creating your Schema and Model as well as accessing them?

I've tried every fix in the thread. With the warning, everything works fine. Once I try `useMongoClient: true` or `mongoose.createConnection` it gets rid of the warning and I can connect to the db, however any code that attempts to access the model fails to trigger. There are no errors, the code just never runs.

**Comment by FinalDes** [comment]
@iamdubx 
just missing the typings, so when run with "noImplicitAny": true , will not have errot 
[example code](https://github.com/FinalDes/graphql-typescript/blob/develop/src/db/db.ts)

**Comment by iamdubx** [comment]
@FinalDes what are you referring to? What missing typings? I have the same code above and have an error.

**Comment by kevinwang309** [comment]
I have encountered the same problem with @Snow-Okami 

1. `const mongoDB = mongoose.connect(config.database, {
useMongoClient: true
});` 
This gets rid of the warning and connects to the db. However, everthing that attempts to access the model fails (get/post). No errors return.
2. `const mongoDB = mongoose.connect(config.database, function(){
useMongoClient: true
});` 
And this gives those two warnings, but everything other than that works fine.

**Comment by Nikunjksanghavi** [comment]
Please at least give backward compatiblity our production build server fails.
feature updation is good but don't change method signature or call.

**Comment by t1bb4r** [comment]
@Nikunjksanghavi It's only a warning, nothing should be failing.

**Comment by FinalDes** [comment]
@iamdubx what kind of error?

**Comment by iamdubx** [comment]
@FinalDes deprecation error, what else we are talking about in this topic?

**Comment by shams-ali** [comment]
This worked for me:

```
const MONGO_URI = process.env.MONGO_URI
const express = require('express');
const app = express();
const mongoose = require('mongoose');
mongoose.Promise = require('bluebird');

// connect to mongo
mongoose.connect(MONGO_URI, {useMongoClient: true})
  .then(({db: {databaseName}}) => console.log(`Connected to ${databaseName}`))
  .catch(err => console.error(err));

const poemSchema = mongoose.Schema({
    name: String,
    text: String
});

const Poem = mongoose.model('Poem', poemSchema);

const insertPoem = poem =>
  (new Poem(poem)).save( (err, newPoem) => 
    console.log(err || newPoem));

insertPoem({name: 'poemName', text: 'this is a poem'})
```

**Comment by saurabh147sharma** [comment]
// DB connection
var mongoose = require('mongoose');
mongoose.Promise = global.Promise;
mongoose.connect('mongodb://localhost/login_register', {
    useMongoClient: true
})
.then(() => console.log('connection succesful'))
.catch((err) => console.error(err));
// DB connection end

**Comment by iamdubx** [comment]
This doesn't work for me. I still have console warning. And no clear explanation provide. 

Mongoose is horrible. Also worst documentation site that I saw in libraries that so popular. What a shame!

**Comment by iamdubx** [comment]
Now I am thinking to switching to RethinkDB. Linux Foundation - this is serious. 
Also Postgre awesome. 

Mongo - very, very shame.

**Comment by shams-ali** [comment]
@iamdubx did you try my solution above? Could you paste the warning on this? Maybe I could help out :)

**Comment by FinalDes** [comment]
@iamdubx can post your warning message?

**Comment by iamdubx** [comment]
@FinalDes @shams-ali I posted before, it's the same deprecation error as topic has. The one this topic is about! 

```
(node:3304) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`. See http://mongoosejs.com/docs/connections.html#use-mongo-client
Server listening on port: 7777
Db.prototype.authenticate method will no longer be available in the next major release 3.x as MongoDB 3.6 will only allow auth against users in the admin db and will no longer allow multiple credentials on a socket. Please authenticate using MongoClient.connect with auth credentials.
```

**Comment by iamdubx** [comment]
@FinalDes @shams-ali I posted before, it's the same deprecation error as topic has. The one this topic is about! 

```
(node:3304) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`. See http://mongoosejs.com/docs/connections.html#use-mongo-client
Server listening on port: 7777
Db.prototype.authenticate method will no longer be available in the next major release 3.x as MongoDB 3.6 will only allow auth against users in the admin db and will no longer allow multiple credentials on a socket. Please authenticate using MongoClient.connect with auth credentials.
```

**Comment by shams-ali** [comment]
@iamdubx interesting, that's the same warning I had and the above solution I posted seemed to fix it for me. Could you paste your updated code that isn't working?

**Comment by afoke** [comment]
This is what worked for me:

// ES6 promises
mongoose.Promise = Promise;

// mongodb connection
mongoose.connect("mongodb://localhost:27017/iptvepg", {
  useMongoClient: true,
  promiseLibrary: global.Promise
});

var db = mongoose.connection;

// mongo error
db.on('error', console.error.bind(console, 'connection error:'));

// mongodb connection open
db.once('open', () => {
  console.log(`Connected to Mongo at: ${new Date()}`)
});

**Comment by FinalDes** [comment]
@afoke is works thanks

**Comment by afoke** [comment]
@FinalDes You're welcome mate.

**Comment by afoke** [comment]
@FinalDes You're welcome mate.

**Comment by betafcc** [comment]
This is my new workflow trying to be compliant with the warnings and deprecations:

    // db.js
    const model = require('./model');
    const mongoose = require('mongoose');
    mongoose.Promise = global.Promise;


    async function db(uri) {
      const con = await mongoose
        .createConnection(uri, {
          useMongoClient: true,
        });

      const {User} = model(con);

      // do something with User
      ...

      // the interface for the server
      return {
        create: params => User.create(params),
        ...
      };
    }


    module.exports = db;


And now the models module returns a function of connection

    // model.js
    const {Schema} = require('mongoose');


    const UserSchema = new Schema({
      name: String,
      age: Number,
    });

    // Other Schemas
    ...

    module.exports = con => ({
      User: con.model('user', UserSchema),
      ...
    });

Any official guidelines on the subject?

**Comment by chrisdothtml** [comment]
@iamdubx Sounds like you don't understand the difference between a **warning** and an **error**. 

The ***warning*** in question does not break anything, but ***warns*** the user of something that will be removed in the next major release.

> Mongoose is horrible.

So keep in mind, you're having this reaction to the developers warning you ahead of time before they remove a feature. I'd also urge you to remember that you're using this free of charge, so maybe give the developers the benefit of the doubt before you throw a tantrum

**Comment by bertolo1988** [comment]
@chrisdothtml 

> The warning in question does not break anything, but warns the user of something that will be removed in the next major release.

It does pollute tests and console which is mildly annoying.

> Mongoose is horrible. Also worst documentation site that I saw in libraries that so popular. What a shame!

He is being impolite which is also mildly annoying but he makes a very valid point in regards to Mongoose documentation. It is not the worst documentation ever but it could really be improved.


The fact that mongoose forwards mongo client API makes inside his own API makes documentation confusing.

A possible solution for this would be:
* stop doing it
* remove the documentation
* provide a way to share connection between the official client and Mongoose

**Comment by jameskentTX** [comment]
I had this error, but it disappeared after I added {useMongoClient: true} option.
I use Debian 9, testing version, with MongoDb ver. 3.2.11 and mongoose ver. 4.10.4.
I have 'Connected to MongoDB' (view code below) message, everything works perfectly.

(<any>mongoose).Promise = global.Promise;
mongoose.connect(process.env.MONGODB_URI, {useMongoClient: true});
const db = mongoose.connection;
db.on('error', console.error.bind(console, 'connection error:'));
db.once('open', () => {
  console.log('Connected to MongoDB');
  setRoutes(app);
  app.get('/*', function(req, res) {
    res.sendFile(path.join(__dirname, '../public/index.html'));
  });
  app.listen(app.get('port'), () => {
    console.log('MyApp listening on port ' + app.get('port'));
  });
});

**Comment by karanpepi** [comment]
Just uninstall the current mongoose version which you have and install the lower version  `npm install mongoose@4.10.8 --save` . No need to do anything else this will definately work for you.

**Comment by systemsky** [comment]
`mongoose.connect(database.host, {
  useMongoClient: true
});`

This solved my problem. But I had another warning coming up after I made this changes.

DeprecationWarning: Mongoose: mpromise (mongoose's default promise library) is deprecated, plug in your own promise library instead: http://mongoosejs.com/docs/promises.html

**Comment by bricss** [comment]
This code will solve all deprecation warnings:
```javascript
mongoose.Promise = global.Promise;
mongoose.connect(uri, {
  keepAlive: true,
  reconnectTries: Number.MAX_VALUE,
  socketTimeoutMS: 0,
  useMongoClient: true
});
````
More info by the link -> http://mongoosejs.com/docs/connections.html#use-mongo-client

**Comment by artnicbr** [comment]
@bricss Fantastic! Thanks!

**Comment by kranz** [comment]
@bricss You made my day.

**Comment by jpetitte** [comment]
someone earned a hero badge...

**Comment by systemsky** [comment]
@bricss : Nicely done Bricks! It worked for me even though I'm using a really old version of mongodb in my local :+1:

**Comment by CodebyOmar** [comment]
@bricss Awesowe! But you might consider putting a smaller number for "reconnectTries". You don't want to spend the whole day reconnecting if there is an issue with your DB.

**Comment by nikkwong** [comment]
@afoke 's solution worked for me, only when i explicitly declare the db in the connection string i.e. ```mongodb://localhost:27017/test``` but not ```mongodb://localhost:27017```

**Comment by aprisniak** [comment]
May I create connection to 2 DB's without this warning? I need to use my models with different DB's.

**Comment by moolsbytheway** [comment]
Using mongoose v^4.11.6 , when added `useMongoClient` the warning dispeared.

`  const uri = "http://blablabla.blo/blaDB";
   mongoose.Promise = global.Promise;
    mongoose.connection.on('error', (err) => {
        console.error(`Mongoose connection error: ${err}`);
        process.exit(1);
    });
    mongoose.connect(uri, {useMongoClient: true});

    // load models
    require('./model1');
    // ...
    require('./modelN');`

**Comment by amado-saladino** [comment]
connecting to database in this way solves the problem:
mongoose.connect(url,{user:'username',pass:'mypassword',useMongoClient:true});

**Comment by iroshanAV** [comment]
Will these errors cause problems , for example in security?

**Comment by RxAssim** [comment]
No more "deprecated" warning pops up if I use this syntax 🎉

```
mongoose.connection.openUri('mongodb://localhost/test')
  .once('open', () => console.log('Good to go !'))
  .on('error', (error) => {
    console.warn('Warning', error);
  });
```

**Comment by mitesh-solulab** [comment]
I faced the same deprecation warning as below:
(node:2300) DeprecationWarning: `open()` is deprecated in mongoose >= 4.11.0, use `openUri()` instead, or set the `useMongoClient` option if using `connect()` or `createConnection()`. See http://mongoosejs.com/docs/connections.html#use-mongo-client

**Solution**

I tried the following and it works fine. Gives no deprecation warnings.

mongoose.connect('mongodb://127.0.0.1:27017/your-database-name', { useMongoClient: true, promiseLibrary: global.Promise });

Hope this will help everyone facing this issue.

**Comment by MichFe** [comment]
I have the same issue but when trying to connect to Mongo Atlas service it just not connect with {useMongoClient:true} option and without it the connection is no longer executed.

Anyone with the same issue?

**Comment by mominbutt12** [comment]
I had solved my problem by just change the mongoose.connect to mongoose.createConnection

mongoose.createConnection(config.uri, (err) => {
  if (err) {
    console.log('Could NOT connect to database: ', err);
  } else {
    console.log('Connected to database: ' + config.db);
  }
});

**Comment by stanisdev** [comment]
I solved this problem like that (Mongoose 4.12.0)

```javascript
mongoose.Promise = require('bluebird');
mongoose.connect('mongodb://localhost:27017/books_exchange', {
  useMongoClient: true,
  promiseLibrary: require('bluebird')
}).then(() => {
  var userSchema = new mongoose.Schema({
    name: String,
    password: String
  });
  var User = mongoose.model('User', userSchema);

  var person = new User({
    name: "John",
    password: "qwerty"
  });

  person.save().then(() => {
    console.log('Data saved');
  }).catch(e => {
    console.log(e);
  });
}).catch(e => {
  console.log('Error while DB connecting');
  console.log(e);
});
```
